### ISSUE TITLE
Issues using ogr2ogr with OGC API vector tiles

### ISSUE THREAD

**Issue by jerstlouis** [OSGeo/gdal#7934]
## Expected behavior and actual behavior.

(in the context of [2023-06 OGC API sprint](https://developer.ogc.org/sprints/21/))

This could potentially be divided into 3 separate issues.

1. First we tried simply using ogr2ogr to export to GeoJSON:

`ogr2ogr test.geojson "OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa:AgricultureSrf"`

That would report an error:

`ERROR 1: Layer 'Zoom level 1' does not already exist in the output dataset, and cannot be created by the output driver.`

It seems that this used to work better without that error with GDAL 3.4, but does not work with latest git master branch.

2. Then we tried selecting a particular tile matrix with e.g., `-oo TILEMATRIX=16` option with the OGC API driver for vector tiles:

`ogr2ogr -oo TILEMATRIX=16 test.geojson "OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa:AgricultureSrf"`
Unfortunately, it seems that this is currently restricted to raster tiles, although the driver does not warn that the TILEMATRIX option is ignored.
Since the driver otherwise outputs duplicate data at different generalized levels, it seems that this option would be even more useful for vector tiles.

3. Finally, we tried to convert to GeoPackage instead:

`ogr2ogr test.gpkg "OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa:AgricultureSrf"`

This took a very long time, but did work in the end, with different layers created for each overview level.
Unfortunately, the geometry is not re-joined so the artificial tile boundaries end up stroked in QGIS.
The lower overview levels should probably also not be used, unless using the [(draft) GeoPackage vector tiles extensions](https://gitlab.com/imagemattersllc/ogc-vtp2/-/tree/master/extensions).

![Agricultural](https://github.com/OSGeo/gdal/assets/5140074/4f20726b-3315-4205-9417-d214c8843e7b)

[agricultureSrf.gpkg.tar.gz](https://github.com/OSGeo/gdal/files/11741607/agricultureSrf.gpkg.tar.gz)

## Steps to reproduce the problem.

1. `ogr2ogr test.geojson "OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa:AgricultureSrf"`
2. `ogr2ogr -oo TILEMATRIX=16 test.geojson "OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa:AgricultureSrf"`
3. `ogr2ogr test.gpkg "OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa:AgricultureSrf"`

## Operating system

Linux

## GDAL version and provenance

GDAL 3.8.0dev-db31b50acf, released 2023/06/12 (debug build)

@doublebyte1 @pixalytics @ghobona

**Comment by rouault** [comment]
> 1. First we tried simply using ogr2ogr to export to GeoJSON:
> 
> 
> `ogr2ogr test.geojson "OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa:AgricultureSrf"`
> 
> That would report an error:
> 
> `ERROR 1: Layer 'Zoom level 1' does not already exist in the output dataset, and cannot be created by the output driver.`
> 
> It seems that this used to work better without that error with GDAL 3.4, but does not work with latest git master branch.

GDAL 3.4 didn't detect the vector tiles, so it used the /items  API as documented "In AUTO mode, for vector access, tiles is used if available, and fallback to GeoJSON items otherwise."
If specifying -oo API=ITEMS, then you the same behavior with master.

I'm not totally sure about the rationale for selecting vector tiles rather than GeoJSON, maybe more compact representation? But maybe GeoJSON /items would be better

> Unfortunately, it seems that this is currently restricted to raster tiles, although the driver does not warn that the TILEMATRIX option is ignored. Since the driver otherwise outputs duplicate data at different generalized levels, it seems that this option would be even more useful for vector tiles.

The driver currently exposes the various zoom levels/tile matrices as separate layers, but indeed it could probably just select the highest one and honour TILEMATRIX. Similarly to the ZOOM_LEVEL + ZOOM_LEVEL_AUTO options of the MBTiles (see https://gdal.org/drivers/raster/mbtiles.html)  and now PMTiles driver (https://github.com/OSGeo/gdal/pull/7938) 

> Unfortunately, the geometry is not re-joined so the artificial tile boundaries end up stroked in QGIS.

When reading tiled vectors, this is unfortunate, but this is not something that cannot be solved when exposing vector tiles with the simple feature model, unless having very small datasets. Unioning geometries accross tiles would be a very costly operation that in the worst case could require downloading the whole dataset before returning a single feature. That's where the fundamental differences in the vector tile vs simple feature models show up. Vector tiles are for display only where you use the buffer around them as a trick to hide the discontinuity. But this is something only a display interface that only copes with a limited number of tiles at once can do.
The only resolution I see for that is to add a warning in the documentatoin page, similary to the one of the MVT, MBTiles and PMTiles driver: "The driver will make no effort of stitching together geometries for
linear or polygonal features that overlap several tiles. An application that wishes to eliminate those interrupts could potentially use the CLIP=NO open option to get larger boundaries, and use appropriate clipping graphic primitives to hide those discontinuities."
And expose this CLIP=NO open option in the OGCAPI driver

**Comment by jratike80** [comment]
> Vector tiles are for display only
Ha, @jerstlouis will soon tell that he loves to use MVT tiles for delivering vector data https://github.com/opengeospatial/2D-Tile-Matrix-Set/issues/86#issuecomment-1566255201 :)

**Comment by jerstlouis** [comment]
> GDAL 3.4 didn't detect the vector tiles,

Ah! Thanks, makes sense.

> I'm not totally sure about the rationale for selecting vector tiles rather than GeoJSON, maybe more compact representation? But maybe GeoJSON /items would be better

Use cases would be:
- A server only supporting _OGC API - Tiles_
- Selecting only a clipped portion of the world for large features at high resolution (e.g., continental coastlines / oceans / landmass polygons), since _Features_ doesn't currently support 'clipping'
- Downloading data at a lower resolution (assuming `-oo TILEMATRIX=` option would work)

> rather than GeoJSON,

[_"vector tiles"_](https://en.wikipedia.org/wiki/Vector_tiles) does not imply _Mapbox_ vector tiles. _OGC API - Tiles_ also has a conformance class for _GeoJSON vector tiles_. Trying to clarify that in an upcoming [conceptual model](https://github.com/opengeospatial/2D-Tile-Matrix-Set/issues/86) ;)

> but indeed it could probably just select the highest one and honour TILEMATRIX.

+1 :)

> Vector tiles are for display only 

I am continuously arguing that they are not (@jratike80 [case in point](https://github.com/OSGeo/gdal/issues/7934#issuecomment-1591989007) :)), at least conceptually (and not talking specifically about Mapbox Vector Tiles) (related to a discussion on a [conceptual model for vector tiles](https://github.com/opengeospatial/2D-Tile-Matrix-Set/issues/86) )

> where you use the buffer around them as a trick to hide the discontinuity

In our client, we do not use the buffer around them to _hide_ the discontinuity, I think this is a wrong approach, that breaks 
for example if you use larger stroke. What we do is clip the geometry to the tiles and mark segments artificially created from the tiling as such. (in a simple feature representation, we simply reference feature/polygon/contour and a from-to point index for any artificial segments).

> Unioning geometries accross tiles would be a very costly operation

Having marked artificial segments facilitates unioning the geometry, as you can easily identify the common vertices along those shared edges.

> in the worst case could require downloading the whole dataset before returning a single feature. 

Indeed vector tiles work better for an interface based on an area and resolution of interest than a feature-based interface.
On the plus side, it can handle arbitrarily large features like a high resolution continental coastline or landmass or ocean, which may take up too much memory. If the tiles can be cached on disk, then the features could then be built from those cached vector tiles.

Related to this would be how the GDAL/OGR vector API (and the OGC API driver) could better serve dynamic clients that focuses on a particular area and resolution of interest? It does seem to already work for raster data. But currently the ability to retrieve the features at the right resolution in QGIS does not work (since overviews are returned as different layers) and things work better with e.g., the [Vector Tiles Reader plugin](https://plugins.qgis.org/plugins/vector_tiles_reader/) than they do with GDAL / OGC API driver.

> The only resolution I see for that is to add a warning
> An application that wishes to eliminate those interrupts could potentially use the CLIP=NO open option to get larger boundaries, and use appropriate clipping graphic primitives to hide those discontinuities.

If I were a real GDAL / QGIS user rather than a developer of competing solutions, I would be very unhappy with that proposed resolution :) Happy to brainstorm a better solution, perhaps in Prizren if you'll be there, I still owe you dinner ;)

**Comment by doublebyte1** [comment]
@rouault @jerstlouis if I try to force the server to use vector tiles with `./ogr2ogr -oo API=TILES test.geojson "OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa:AgricultureSrf", I get this response:

ERROR 1: Layer 'Zoom level 1' does not already exist in the output dataset, and cannot be created by the output driver.
ERROR 1: Terminating translation prematurely after failed

However, ogrinfo is able to read the dataset:

./ogrinfo -oo API=ITEMS  "OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa:AgricultureSrf"
INFO: Open of `OGCAPI:https://maps.gnosis.earth/ogcapi/collections/Daraa:AgricultureSrf'
      using driver `OGCAPI' successful.
Metadata:
  TITLE=AgricultureSrf
1: Daraa:AgricultureSrf (title: AgricultureSrf) (Polygon)

I agree that for this particular use case of creating a vector file, it is much better to use features, but what I am trying to track down here is why QGIS is not able to open vector tiles using the ogcapi driver (see screenshot), while it is able to open raster tiles.


![Screenshot from 2023-06-19 02-01-43](https://github.com/OSGeo/gdal/assets/1038897/2cd21ac2-3194-4437-97ed-5bf3f4db8411)

Any ideas on what the problem may be?

I am on GDAL 3.8.0dev-23d309f3f4, released 2023/05/02 (debug build)

**Comment by jerstlouis** [comment]
Thanks @doublebyte1 I think that's OGR not knowing what to do with the different overviews when outputting to a single GeoJSON file, as we discussed above.
Ideally, this would be fixed so that it picks the most detailed zoom level by default, and would also support `-oo TILEMATRIX=` to pick a lower resolution.

As mentioned above, ogr2ogr with `-oo API=ITEMS` will also work using _Features_, as it used to with GDAL 3.4.

I don't know if the QGIS issue is related from the OGR API perspective of not being able to pick a single level or retrieve each of them individually?

> it is much better to use features

_OGC API - Tiles_ with vector tiles should really be able to cover most scenarios where _Features_ can be used, and more (high resolution datasets covering a large spatial area), efficiently.

**Comment by doublebyte1** [comment]
@jerstlouis  I am not sure showing features, when tiles were specifically requested is fair for the user. One thing is to have a default behavior when no parameters are given, but to override options seems a bit misleading.

I think that picking a zoom level (the maximum?), when this is not specified seems a good option. Isn't that what is happening when we request raster tiles?

**Comment by jerstlouis** [comment]
> @jerstlouis I am not sure showing features, when tiles were specifically requested is fair for the user.

I'm not sure what you mean there.
I was not saying that GDAL should use `/items` when `-oo API=TILES` is specified.

Retrieving vector tiles from _OGC API - Tiles_ retrieves _tiled feature data_, so you are requesting portions of features, potentially a generalized version of them if retrieving for lower zoom levels.

> I think that picking a zoom level (the maximum?), when this is not specified seems a good option. Isn't that what is happening when we request raster tiles?

With raster tiles, QGIS / GDAL seems to support retrieving tiles at the correct resolution / scale for the current view.
I believe this functionality that exists for raster data might not yet be part of the GDAL/OGR API for vector data.

**Comment by doublebyte1** [comment]
I think we could start by sorting the zoom level issue on GDAL, before starting to look at QGIS. I created an issue here: https://github.com/OSGeo/gdal/issues/7969

**Comment by jerstlouis** [comment]
@doublebyte1 QGIS uses the GDAL/OGR API, so the problem is likely unrelated to ogr2ogr's ability to pick a tile matrix or not.

ogr2ogr works fine when outputting to a GeoPackage that can store the overviews as multiple layers.

#7969 is more specific, but I think this was the essence of what could easily be fixed about this issue, so it is a bit of a duplicate of this one.

The only other things mentioned in this issue that could be improved are large work items:
- improving GDAL/OGR to support automatically and efficiently re-combining vector tiles and 
- improving the API to allow retrieving vector tiles for the client's current zoom levels, just like with raster tiles.

Probably new "feature requests" issues could be opened for these two things (if one does not already exist), and this issue could then be closed.

**Comment by jerstlouis** [comment]
I have opened the two other separate issues, so closing this issue now that everything here is now covered by:

- #7969
- #7979 
- #7980
